File: src/TON_UPDATES/python_code/randpolicy.py
"""
randpolicy.py - 1:1 Python port of randpolicy.m

Randomized scheduling baseline: pick min(N, M) sources uniformly at random
without replacement, assign each a uniformly random task. No capacity gating.
Average the discounted sum of errors over Titer=100 Monte-Carlo trials.
"""
import numpy as np


def randpolicy(M, N, km, T, B, K, n, c, w, gamma, p, titer=100, verbose=True):
    n = np.asarray(n, dtype=float)
    w = np.asarray(w, dtype=float)
    p = np.asarray(p, dtype=float)

    presultiter = np.zeros(titer)
    for it in range(titer):
        Delta = np.zeros((M, km), dtype=np.int64)
        pavg = np.zeros(K)
        for t in range(K):
            Change = np.zeros((M, km), dtype=bool)

            n_select = min(N, M)
            source = np.random.permutation(M)[:n_select]
            for row in source:
                column = np.random.randint(0, km)  # MATLAB randperm(km, 1)
                Delta[row, column] = 0
                Change[row, column] = True

            for m in range(M):
                for j in range(km):
                    if not Change[m, j]:
                        if Delta[m, j] + 1 > B - 1:
                            Delta[m, j] = B - 1
                        else:
                            Delta[m, j] = Delta[m, j] + 1
                    pavg[t] += (w[m, j] * p[j, Delta[m, j]]) / (km * M)

            if t + 1 < T:
                presultiter[it] += (gamma ** t) * pavg[t]

    presult = float(np.mean(presultiter))
    if verbose:
        print(f"randpolicy presult = {presult}")
    return presult

File: src/TON_UPDATES/matlab_code/randpolicy.m
function presult=randpolicy(M,N,km,T,B,K,n,c,w,gamma,p)
presult=0;
Titer=100;
presultiter=zeros(1,Titer);
for iter=1:Titer
    Delta=ones(M,km); % State initialization
    pavg=zeros(1,K);
for t=1:K
    %% Initialization
    g=zeros(M,km);
    Change=zeros(M,km);
    source=randperm(M,min(N,M));
    i=1;
    for i=1:length(source)
           row=source(i);
            column=randperm(km,1);
            g(row, column)=1;
            Delta(row, column)=1;
            Change(row, column)=1;
    end

    %% update AoI
    for m=1:M
        for j=1:km
            if Change(m,j)==0
                if Delta(m, j)+1>B
                    Delta(m, j)=B;
                else
                    Delta(m, j)=Delta(m, j)+1;
                end
            end
           pavg(t)=pavg(t)+(w(m, j)*p(j,Delta(m,j)))/(km*M);
        end
    end
   
    if t<T
    presultiter(iter)=presultiter(iter)+gamma^(t-1)*pavg(t);
    end
end
end
presult=mean(presultiter)
end